src/VisualStudio/Core/Def/TaskList/ExternalErrorDiagnosticUpdateSource.cs

```diff
@@ -139,385 +110,160 @@ public ImmutableArray<DiagnosticData> GetBuildErrors()
 
-        // Raise events for build only errors
-        using var argsBuilder = TemporaryArray<DiagnosticsUpdatedArgs>.Empty;
-        var buildErrors = GetBuildErrors().Except(allLiveErrors).GroupBy(k => k.DocumentId);
-        foreach (var group in buildErrors)
+        if (projectId is not null)
         {
-            cancellationToken.ThrowIfCancellationRequested();
-
-            if (group.Key == null)
-            {
-                foreach (var projectGroup in group.GroupBy(g => g.ProjectId))
-                {
-                    Contract.ThrowIfNull(projectGroup.Key);
-                    argsBuilder.Add(CreateArgsToReportBuildErrors(projectGroup.Key, solution, projectGroup.ToImmutableArray()));
-                }
-
-                continue;
-            }
-
-            argsBuilder.Add(CreateArgsToReportBuildErrors(group.Key, solution, group.ToImmutableArray()));
+            await diagnosticManagerService.ClearDiagnosticsAsync(projectId.Id.ToString(), cancellationToken).ConfigureAwait(false);
         }
-
-        ProcessAndRaiseDiagnosticsUpdated(argsBuilder.ToImmutableAndClear());
-
-        // Report pending live errors
-        return diagnosticService.SynchronizeWithBuildAsync(_workspace, pendingLiveErrorsToSync, onBuildCompleted: true, cancellationToken);
-    }
-
-    private static DiagnosticsUpdatedArgs CreateArgsToReportBuildErrors<T>(T item, Solution solution, ImmutableArray<DiagnosticData> buildErrors)
-    {
-        if (item is ProjectId projectId)
+        else
         {
-            return CreateDiagnosticsCreatedArgs(projectId, solution, projectId, documentId: null, buildErrors);
+            await diagnosticManagerService.ClearAllDiagnosticsAsync(cancellationToken).ConfigureAwait(false);
         }
-
-        RoslynDebug.Assert(item is DocumentId);
-        var documentId = (DocumentId)(object)item;
-        return CreateDiagnosticsCreatedArgs(documentId, solution, documentId.ProjectId, documentId, buildErrors);
     }
 
-    private static void AddArgsToClearBuildOnlyProjectErrors(ref TemporaryArray<DiagnosticsUpdatedArgs> builder, Solution solution, ProjectId? projectId)
+    private async ValueTask ProcessDiagnosticsReportAsync(ProjectId projectId, Guid projectHierarchyGuid, ImmutableArray<DiagnosticData> diagnostics, InProgressState state, CancellationToken cancellationToken)
     {
-        // Remove all project errors
-        builder.Add(CreateDiagnosticsRemovedArgs(projectId, solution, projectId, documentId: null));
+        var diagnosticManagerService = await GetOrCreateDiagnosticManagerAsync(cancellationToken).ConfigureAwait(false);
 
-        var project = solution.GetProject(projectId);
-        if (project == null)
+        using var _ = ArrayBuilder<DiagnosticCollection>.GetInstance(out var collections);
+        var groupedDiagnostics = diagnostics.GroupBy(d => d.DataLocation.UnmappedFileSpan.Path);
```

Comment:
note: this will group across the same document linked into multiple projects.  if that's the actual intent here, can you doc?